src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DraftReferralService.kt

```diff
@@ -487,6 +498,78 @@ class DraftReferralService(
     }
   }
 
+  private fun verifyCustodyLocationEntries(referral: Referral) {
+    val eventName = "CustodyLocationVerification"
+    var eventProperties: Map<String, String> = emptyMap()
+    val compare = { a: String?, b: String? ->
+      when {
+        a == null || b == null -> MatchType.UNDETERMINED.name
+        a == b -> MatchType.MATCH.name
+        a != b -> MatchType.NO_MATCH.name
+        else -> MatchType.UNDETERMINED.name
+      }
+    }
+
+    val offenderNomsId = communityAPIOffenderService.getOffenderIdentifiers(referral.serviceUserCRN)
+      .onErrorResume(WebClientResponseException::class.java) { e ->
+        eventProperties = mapOf(
+          "result" to "error",
+          "errorStatusCode" to e.rawStatusCode.toString(),
+          "errorReason" to e.statusCode.reasonPhrase
+        )
+        Mono.empty()
+      }
+      .block()?.primaryIdentifiers?.nomsNumber
+
+    offenderNomsId?.let {
+      val prisoner = prisonerOffenderSearchService.getPrisonerById(it)
+        .onErrorResume(WebClientResponseException::class.java) { e ->
+          eventProperties = mapOf(
+            "result" to "error",
+            "errorStatusCode" to e.rawStatusCode.toString(),
+            "errorReason" to e.statusCode.reasonPhrase
+          )
+          Mono.empty()
+        }
+        .block()
+
+      prisoner?.let {
+        val referralLocation = referralLocationRepository.findByReferralId(referral.id)
+        referralLocation?.let {
+          with(prisoner) {
+            it.nomisPrisonId = prisonId
+            it.nomisReleaseDate = releaseDate
+            it.nomisConfirmedReleaseDate = confirmedReleaseDate
+            it.nomisNonDtoReleaseDate = nonDtoReleaseDate
+            it.nomisAutomaticReleaseDate = automaticReleaseDate
+            it.nomisPostRecallReleaseDate = postRecallReleaseDate
+            it.nomisConditionalReleaseDate = conditionalReleaseDate
+            it.nomisActualParoleDate = actualParoleDate
+            it.nomisDischargeDate = dischargeDate
+
+            eventProperties = mapOf(
+              "result" to "success",
+              "prisonId" to compare(it.nomisPrisonId as String, it.prisonId),
+              "releaseDate" to compare(it.nomisReleaseDate.toString(), it.expectedReleaseDate.toString())
+            )
+          }
+          referralLocationRepository.save(referralLocation)
+        }
+      } ?: run {
+        if (eventProperties.isEmpty()) eventProperties = mapOf(
+          "result" to "undetermined",
+          "reason" to "Prisoner not found for NOMS id: $offenderNomsId"
+        )
+      }
+    } ?: run {
+      if (eventProperties.isEmpty()) eventProperties = mapOf(
+        "result" to "undetermined",
+        "reason" to "NOMS id not found for CRN: ${referral.serviceUserCRN}"
+      )
+    }
+    telemetryClient.trackEvent(eventName, eventProperties, null)
```

Comment:
truth table
| noms lookup | prison search | location matching | dates | CustodyLocationVerification properties
| -- | -- |-- |-- | -- 
| missing data | | | | result: noms not found, crn: $crn
| 4xx, 5xx, network error | | | | result: noms lookup failed, exception: $message, crn: $crn
| / | 4xx, 5xx, network error | | | result: prison search failed, exception: $message, crn: $crn
| / | / | do matching | no need to compare anything | result: match result, crn: $crn